torch/_export/non_strict_utils.py

```diff
@@ -0,0 +1,145 @@
+from torch._guards import Source
+from torch._subclasses.fake_tensor import FakeTensorMode
+from torch.export import Constraint
+from torch.fx.experimental.symbolic_shapes import (
+    DimDynamic,
+    EqualityConstraint,
+    ShapeEnv,
+    StatelessSymbolicContext,
+)
+
+
+def fakify(mode, t, tensor_constraints, source, dim_name_to_sources):
+    n_dims = len(t.shape)
+    symbolic_context = StatelessSymbolicContext(
+        dynamic_sizes=[DimDynamic.STATIC] * n_dims,
+        constraint_sizes=[None] * n_dims,
+    )
+    t_id = id(t)
+    if t_id in tensor_constraints:
+        for i, constraint in tensor_constraints[t_id].items():
+            symbolic_context.constraint_sizes[i] = constraint.constraint_range
+            symbolic_context.dynamic_sizes[i] = DimDynamic.DYNAMIC
+            src = TensorPropertySource(base=source, prop=TensorProperty.SIZE, idx=i)
+            dim_name_to_sources[constraint.debug_name].append(src)
+            mode.shape_env.source_name_to_debug_name[src.name()] = constraint.debug_name
+    fake = mode.from_tensor(t, source=source, symbolic_context=symbolic_context)
+    mode.shape_env.tracked_fakes.append(
+        TrackedFake(fake, source, symbolic_context.constraint_sizes)
+    )
+    return fake
+
+
+def fake_tree(mode, arg, tensor_constraints, source, dim_name_to_sources):
+    if isinstance(arg, list):
+        return [
+            fake_tree(
+                mode,
+                arg,
+                tensor_constraints,
+                GetItemSource(source, i),
+                dim_name_to_sources,
+            )
+            for i, arg in enumerate(arg)
+        ]
+
+    elif isinstance(arg, dict):
+        return {
+            k: fake_tree(
+                mode,
+                arg,
+                tensor_constraints,
+                GetItemSource(source, k),
+                dim_name_to_sources,
+            )
+            for k, arg in arg.items()
+        }
+
+    else:
+        return fakify(mode, arg, tensor_constraints, source, dim_name_to_sources)
+
+
+def make_fake_inputs(nn_module, args, constraints):
+    tensor_constraints: Dict[int, Dict[int, Constraint]] = defaultdict(dict)
+    for constraint in constraints:
+        tensor_constraints[constraint.t_id][constraint.dim] = constraint
+        if constraint.shared is not None:
+            tensor_constraints[constraint.shared.t_id][constraint.shared.dim] = constraint
+
+    code = nn_module.forward.__code__
+    co_fields = {
+        "co_name": code.co_name,
+        "co_filename": code.co_filename,
+        "co_firstlineno": code.co_firstlineno,
+    }
+    with FakeTensorMode(
+        shape_env=ShapeEnv(tracked_fakes=[], co_fields=co_fields)
+    ) as fake_mode:
+        params = inspect.signature(nn_module.forward).parameters
+        dim_name_to_sources: Dict[str, List[Source]] = defaultdict(list)
+        fake_args = tuple(
+            fake_tree(
```

Comment:
nit: Why can't we use pytree here?

Reply:
Yeah that would be nice, but unfortunately cannot. We're building sources too as we descend. (I can check if there's a pytree util that can do additional bookkeeping as we descend...)